ocaml/xapi/xapi_database_backup.ml

```diff
@@ -0,0 +1,175 @@
+
+type count = {tblname:string; count: int; del:int;}
+type pair = {tbl:string; row:string;}
+
+let empty_schema = Schema.Value.unmarshal String ""
+let empty_field = {fldname=""; stat={created=0L;modified=0L;deleted=0L;}; value=empty_schema;}
+let empty_row   = {objref =""; stat={created=0L;modified=0L;deleted=0L;}; value=[empty_field];}
+
+exception Content_length_required
+
+let delta_to_string (d:delta) = Jsonrpc.to_string (rpc_of_delta d)
+
+(* Master Functions *)
+
+let get_del_from_table_name token ts table =
+  Table.fold_over_deleted token (fun key {created; modified; deleted} acc ->
+      ({key; table; stat={created;modified;deleted}})::acc) (TableSet.find table ts) []
+
+let get_deleted ~__context ~token db =
+  let tables =
+    let all =
+      let objs = List.filter (fun x -> x.Datamodel_types.gen_events) (Dm_api.objects_of_api Datamodel.all_api) in
+      List.map (fun x->x.Datamodel_types.name) objs in
+    List.filter (fun table -> true) all in
+  List.concat (List.map (get_del_from_table_name token (Database.tableset db)) tables)
+
+let check_for_updates token db =
+  let ts = Database.tableset db in
+  let handle_obj obj = Row.fold_over_recent token (fun fldname {created; modified; deleted} value acc
+                                                    -> {fldname; stat={created;modified;deleted}; value=value}::acc) obj [] in
+  let handle_table table = Table.fold_over_recent token (fun objref {created; modified; deleted} value acc
+                                                          -> {objref; stat={created;modified;deleted}; value=(handle_obj value)}::acc) table [] in
+  TableSet.fold_over_recent token (fun tblname {created; modified; deleted} value acc
+                                    -> {tblname; stat={created;modified;deleted}; value=(handle_table value)}::acc) ts []
+
+let get_delta ~__context ~token =
+  let db = Db_ref.get_database (Context.database_of __context) in
+  let db_access =
+    fun () -> Mutex.execute Db_lock.dbcache_mutex (fun () ->
+        {fresh_token=Manifest.generation (Database.manifest db);
+         tables=(check_for_updates token db);
+         deletes=(get_deleted __context token db)})
+  in
+  let _dbreply = db_access () in
+  _dbreply
+
+let handler (req: Request.t) s _ =
+  req.Http.Request.close <- true;
+  let token = Generation.of_string (snd (List.find (fun (x,y) -> x = "token") req.query)) in
+  Xapi_http.with_context "Database backup" req s (fun __context ->
+      let d = get_delta __context token in
+      let str_update = (Jsonrpc.to_string (rpc_of_delta d)) in
+      Http_svr.response_str req s str_update
+    )
+
+(* Slave Functions *)
+
+let send_notification (c:Context.t) reason p =
+  let record = (Eventgen.find_get_record p.tbl ~__context:c ~self:p.row) () in
+  begin match record with
+    | None -> debug "Could not send %s event for %s in %s table" reason p.row p.tbl;
+    | Some record -> debug "Sending %s event for %s object in %s table" reason p.row p.tbl;
```

Comment:
This is probably going to result in lots and lots of log lines? This may be handy for dev testing, but not in the final version of the patch.